packages/autocomplete-core/src/reshape.ts

```diff
@@ -0,0 +1,55 @@
+import { flatten } from '@algolia/autocomplete-shared';
+
+import {
+  AutocompleteCollection,
+  AutocompleteReshapeSourcesBySourceId,
+  AutocompleteState,
+  BaseItem,
+  InternalAutocompleteOptions,
+} from './types';
+
+type ReshapeParams<TItem extends BaseItem> = {
+  collections: Array<AutocompleteCollection<any>>;
+  props: InternalAutocompleteOptions<TItem>;
+  state: AutocompleteState<TItem>;
+};
+
+export function reshape<TItem extends BaseItem>({
+  collections,
+  props,
+  state,
+}: ReshapeParams<TItem>) {
+  // Sources are grouped by `sourceId` to conveniently pick them via destructuring.
+  // Example: `const { recentSearchesPlugin } = sourcesBySourceId`
+  const sourcesBySourceId = collections.reduce<
```

Comment:
could reuse `groupBy` 🤔

Reply:
The `groupBy` reshape function is dedicated to Autocomplete sources so it's a bit less trivial than that one.